awscli/customizations/s3/results.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import logging
+from collections import namedtuple
+
+from s3transfer.subscribers import BaseSubscriber
+
+from awscli.customizations.s3.utils import relative_path
+
+
+LOGGER = logging.getLogger(__name__)
+
+QueuedResult = namedtuple(
+    'QueuedResult',
+    ['transfer_type', 'src', 'dest', 'total_transfer_size']
+)
+ProgressResult = namedtuple(
+    'ProgressResult',
+    ['transfer_type', 'src', 'dest', 'bytes_transferred',
+     'total_transfer_size']
+)
+SuccessResult = namedtuple(
+    'SuccessResult', ['transfer_type', 'src', 'dest']
+)
+FailureResult = namedtuple(
+    'FailureResult', ['transfer_type', 'src', 'dest', 'exception']
+)
+CommandResult = namedtuple(
+    'CommandResult', ['num_tasks_failed', 'num_tasks_warned'])
+
+
+class BaseResultSubscriber(BaseSubscriber):
+    TRANSFER_TYPE = None
+
+    def __init__(self, result_queue):
+        """Subscriber to send result notifications during transfer process
+
+        :param result_queue: The queue to place results to be processed later
+            on.
+        """
+        self._result_queue = result_queue
+        self._transfer_type = None
+
+    def on_queued(self, future, **kwargs):
+        src, dest = self._get_src_dest(future)
+        queued_result = QueuedResult(
+            self.TRANSFER_TYPE, src, dest, future.meta.size)
+        self._result_queue.put(queued_result)
+
+    def on_progress(self, future, bytes_transferred, **kwargs):
```

Comment:
Yep it is new. And definitely we can make more detailed progress. For the ResultPrinter since there is no notion of "parts", I am planning to print progress in terms of human readable bytes.